docs/coding-guidelines.md

```diff
@@ -0,0 +1,269 @@
+    RootBehaviour <|-- IdentifyBehaviour
+    RootBehaviour <|-- KademliaBehaviour
+
+    Swarm : poll()
+    RootBehaviour : poll()
+    ConnectionPool : poll()
+    Transport : poll()
+    PingBehaviour : poll()
+    IdentifyBehaviour : poll()
+    KademliaBehaviour : poll()
+    @enduml
+    ```
+</details>
+
+```
+                                         ,------.
+                                         |Swarm |
+                                         |------|
+                                         |poll()|
+                                         `------'
+                                             |
+                                             |
+                   ,-------------.   ,--------------.   ,---------.
+                   |RootBehaviour|   |ConnectionPool|   |Transport|
+                   |-------------|   |--------------|   |---------|
+                   |poll()       |   |poll()        |   |poll()   |
+                   `-------------'   `--------------'   `---------'
+                           |
+,-------------.  ,-----------------.   ,-----------------.
+|PingBehaviour|  |IdentifyBehaviour|   |KademliaBehaviour|
+|-------------|  |-----------------|   |-----------------|
+|poll()       |  |poll()           |   |poll()           |
+`-------------'  `-----------------'   `-----------------'
+```
+
+Using hierarchical state machines is a deliberate choice throughout the
+rust-libp2p code base. It makes reasoning about control and data flow easy and
+works well with Rust's `Future` model. The archetecture pattern of
+hierarchical state machines should be used wherever possible.
+
+### Conventions for `poll` implementations
+
+The `poll` method of a single state machine can be complex especially when that
+state machine itself `poll`s many child state machines. The patterns shown below
+have proven useful and should be followed across the code base.
+
+``` rust
+fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>{
+  loop {
+    match self.child_1.poll(cx) {
+      // The child made progress.
+      Poll::Ready(_) => {
+        // Either return an event to the parent:
+        return Poll::Ready(todo!());
+        // or `continue`, thus polling `child_1` again. `child_1` can potentially make more progress:
+        continue
+        // but NEVER move to the next child if the current child made progress. Given
+        // that the current child might be able to make more progress, it did not yet
+        // register the waker in order for the root task to be woken up later on. Moving
+        // on to the next child might result in the larger `Future` task to stall as it
+        // assumes that there is no more progress to be made.
+      }
+
+      // The child did not make progress. It has registered the waker for a
+      // later wake up. Proceed with the other childs.
+      Poll::Ready(_) => {
+    }
+
+    match self.child_2.poll(cx) {
+      // As above.
+    }
```

Comment:
I think it is important to mention that using continue in here will go back to child1 which could make progress again.

There is a subtle difference between looping over the entire thing and exhausting each child until it returns Pending. We want the former, not the latter.

Reply:
Agreed, though not sure how to put it into words.

There is this line already:

``` rust
        // or `continue`, thus polling `child_1` again. `child_1` can potentially make more progress:
        continue
```

Could you make a suggestion @thomaseizinger ?

Reply:
I'd suggest to just fill in the `match` here too and make a comment next to the `continue` of the `Poll::Pending` branch.

Reply:
> I'd suggest to just fill in the `match` here too

As in the same as the above? What would be the benefit of that? Isn't that just duplicating the same comments once more?

Reply:
I wouldn't duplicate the entire explanation but perhaps have a complete example of the control flow and make one comment at the bottom that highlights that we always `continue` to go back to the top. Even though it should be clear from the written text, I find that seeing the entire control flow is better.

Maybe also build in the case where the event from `child2` is dispatched to `child1`?

Reply:
I added https://github.com/libp2p/rust-libp2p/pull/2780/commits/ed4d6330487e5a0e8bf8dc61bc9e14367b32c5c8. Let me know what you think.